File: medium/092_Reverse_Linked_List_II/reverse.c
#include <stddef.h>
#include "../../common/list.h"

static void
reverse_util (list_node_t *start, list_node_t *end)
{
    list_node_t *prev, *cur, *next;

    prev = NULL;
    cur = start;
    for (;;) {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
        if (prev == end) {
            break;
        }
    }
    return;
}

list_node_t *
reverseBetween (list_node_t *head, int m, int n)
{
    int i = 1;
    list_node_t *cur, *next, *start = NULL, **pprev;

    if (head == NULL) {
        return NULL;
    }
    if (m == n) {
        return head;
    }

    pprev = &head;
    for (cur = head; cur; cur = cur->next, i++) {
        if (i == m) {
            start = cur;
        } else if (i < m) {
            pprev = &cur->next;
        } else if (i == n) {
            next = cur->next;
            reverse_util(start, cur);
            *pprev = cur;
            start->next = next;
            break;
        }
    }

    return head;
}

#include <stdio.h>

static void
test_reverse_list (int *a, int len, int m, int n)
{
    list_node_t *head, *p;

    head = make_list(a, len);
    if (head == NULL) {
        printf("Fail to make list.\n");
        return;
    }

    p = head;

    printf("Test reverse %d, %d\n", m, n);
    printf("Before reverse: ");
    dump_list(head);
    head = reverseBetween(head, m, n);
    printf("After reverse: ");
    dump_list(head);
    printf("\n");

    free_list(p);
    return;
}

#define ARRAY_LEN(a) (sizeof(a)/sizeof((a)[0]))

int main (void)
{
    int a0[] = {1, 2, 3, 4, 5, 6, 7};
        
    test_reverse_list(a0, ARRAY_LEN(a0), 1, 7);
    test_reverse_list(a0, ARRAY_LEN(a0), 1, 6);
    test_reverse_list(a0, ARRAY_LEN(a0), 1, 2);
    test_reverse_list(a0, ARRAY_LEN(a0), 2, 7);
    test_reverse_list(a0, ARRAY_LEN(a0), 6, 7);
    test_reverse_list(a0, ARRAY_LEN(a0), 2, 3);
    test_reverse_list(a0, ARRAY_LEN(a0), 5, 5);
    return 0;
}

File: medium/092_Reverse_Linked_List_II/reverse.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, m int, n int) *ListNode {
	if m == n {
		return head
	}
	var dummy = &ListNode{Next: head}
	var prev, next, cur, first, tail *ListNode
	prev = dummy
	i := 1
	for cur = head; cur != nil; {
		next = cur.Next
		if i == m {
			first, tail = cur, prev
		} else if i > m {
			cur.Next = prev
			if i == n {
				first.Next, tail.Next = next, cur
				break
			}
		}
		prev, cur = cur, next
		i++
	}
	return dummy.Next
}

func numArrayToList(nums []int) *ListNode {
	var l, prev *ListNode
	for _, n := range nums {
		node := &ListNode{Val: n}
		if prev == nil {
			l = node
		} else {
			prev.Next = node
		}
		prev = node
	}
	return l
}

func dumpList(head *ListNode) {
	if head == nil {
		fmt.Println("Nil")
		return
	}
	for p := head; p != nil; p = p.Next {
		if p == head {
			fmt.Printf("%d", p.Val)
		} else {
			fmt.Printf(" -> %d", p.Val)
		}
	}
	fmt.Println()
}

func testReverse(nums []int, m, n int) {
	head := numArrayToList(nums)
	fmt.Printf("Before m %d, n %d: ", m, n)
	dumpList(head)
	head = reverseBetween(head, m, n)
	fmt.Printf("After m %d, n %d:  ", m, n)
	dumpList(head)
}

func main() {
	testReverse([]int{1, 2, 3, 4, 5}, 1, 1)
	testReverse([]int{1, 2, 3, 4, 5}, 1, 2)
	testReverse([]int{1, 2, 3, 4, 5}, 1, 5)
	testReverse([]int{1, 2, 3, 4, 5}, 2, 3)
	testReverse([]int{1, 2, 3, 4, 5}, 2, 5)
	testReverse([]int{1, 2, 3, 4, 5}, 4, 5)
	testReverse([]int{1, 2, 3, 4, 5}, 5, 5)
}
